src/ferny/ssh_askpass.py

```diff
@@ -0,0 +1,181 @@
+        return await responder.do_prompt(self)
+
+
+class SSHAskpassPrompt(AskpassPrompt):
+    ANSWERS: ClassVar[Optional[Sequence[str]]] = None
+    EXTRA_PATTERNS: ClassVar[Sequence[str]] = ()
+    PATTERN: ClassVar[str]
+
+    def __init__(self, matches: Optional[Dict[str, str]] = None) -> None:
+        if matches is not None:
+            self.__dict__.update(matches)
+
+
+# Specific prompts
+HELPERS = {
+    "%{algorithm}": r"(?P<algorithm>\b[-\w]*\b)",
+    "%{filename}": r"(?P<filename>.+)",
+    "%{fingerprint}": r"(?P<fingerprint>SHA256:[0-9A-Za-z+/]{43})",
+    "%{hostname}": r"(?P<hostname>[^ @']+)",
+    "%{pkcs11_id}": r"(?P<pkcs11_id>.+)",
+    "%{username}": r"(?P<username>[^ @']+)",
+}
+
+
+class PasswordPrompt(SSHAskpassPrompt):
+    PATTERN = r"%{username}@%{hostname}'s password: "
+    username: Optional[str] = None
+    hostname: Optional[str] = None
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_password_prompt(self)
+
+
+class PassphrasePrompt(SSHAskpassPrompt):
+    PATTERN = r"Enter passphrase for key '%{filename}': "
+    filename: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_passphrase_prompt(self)
+
+
+class FIDOPINPrompt(SSHAskpassPrompt):
+    PATTERN = r"Enter PIN for %{algorithm} key %{filename}: "
+    algorithm: str
+    filename: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_fido_pin_prompt(self)
+
+
+class FIDOUserPresencePrompt(SSHAskpassPrompt):
+    PATTERN = r"Confirm user presence for key %{algorithm} %{fingerprint}\n"
+    ANSWERS = ()
+    algorithm: str
+    fingerprint: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_fido_user_presence_prompt(self)
+
+
+class PKCS11PINPrompt(SSHAskpassPrompt):
+    PATTERN = r"Enter PIN for '%{pkcs11_id}: "
+    pkcs11_id: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_pkcs11_pin_prompt(self)
+
+
+class HostKeyPrompt(SSHAskpassPrompt):
+    PATTERN = r"Are you sure you want to continue connecting \(yes/no(/\[fingerprint\])?\)\? "
+    EXTRA_PATTERNS = [
+        r"%(fingerprint)[.]$",
+        r"^%{algorithm} key fingerprint is",
+        r"^The fingerprint for the %{algorithm} key sent by the remote host is$"
+    ]
+    ANSWERS = ('yes', 'no')
+    algorithm: str
+    fingerprint: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_host_key_prompt(self)
+
+
+def with_helpers(pattern: str) -> str:
+    for name, helper in HELPERS.items():
+        pattern = pattern.replace(name, helper)
+
+    assert '%{' not in pattern
+    return pattern
+
+
+def categorize_ssh_prompt(string: str, stderr: str) -> AskpassPrompt:
```

Comment:
I don't see where `EXTRA_PATTERNS` is ever being looked at. `HostKeyPrompt` defines it, and the base class declares it, but it's otherwise ignored.

Reply:
Probably leftover from a previous approach.  I'll take another look.

Reply:
I added the code for dealing with this properly.  I'm also going to expand the testcases to make sure we read the correct fingerprint, etc.